homeassistant/components/incharge/sensor.py

```diff
@@ -0,0 +1,152 @@
+"""Get status for all stations for InCharge."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import datetime
+import logging
+from typing import Any
+
+from incharge.api import InCharge
+import requests
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, UnitOfEnergy
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import UpdateFailed
+from homeassistant.util import Throttle
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = datetime.timedelta(minutes=5)
+
+
+@dataclass
+class InChargeSensorEntityDescription(SensorEntityDescription):
+    """Describes InCharge sensor entity."""
+
+    precision: int | None = None
+
+
+SENSOR_TYPES: dict[str, InChargeSensorEntityDescription] = {
+    "total_energy_consumption": InChargeSensorEntityDescription(
+        key="total_energy_consumption",
+        name="Total Energy Consumption",
+        precision=2,
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+    )
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the InCharge sensor."""
+    config = {**config_entry.data}
+    username = config[CONF_USERNAME]
+    password = config[CONF_PASSWORD]
+
+    incharge_api = InCharge(username, password)
+    try:
+        response = await hass.async_add_executor_job(incharge_api.get_stations)
+        response.raise_for_status()
+        stations = [station["name"] for station in response.json().get("stations")]
+    except (
+        ConnectionError,
+        requests.exceptions.HTTPError,
+    ) as incharge_connection_error:
+        raise ConfigEntryAuthFailed from incharge_connection_error
```

Comment:
You should be doing this in init and then setting hass.data[entry.entry_id] to your coordinator or whatever object you use to hold your api 

Then you receive the coordinator here from hass.data

Your incharge data obj potentially?

You should only have one instance of an object that handles updating and holds your api though - so you would have to change it

Reply:
I was a bit lost on why the async_setup_entry is both here and in the __init__ file, should all content of this function be merged with the same function in the __init__ file?

Reply:
At the same time I want to make sure that when there are new entities (in this case a new charging station) it is automatically added but looking into this I am not sure if this is the case right now even...

Reply:
So in init, you handle the general setup and you then forward entry setup, so you end up setting up everything that is generic in the init one, and then you "forwards the set up for anything platform specific in the platform async setup

Reply:
Ok agree, I will change it